# Before you start :
   - These exercises are related to the Pivot table and correlation lessons.
   - Keep in mind that you need to use some of the functions you learned in the previous lessons.
   - All datasets are provided in the `your-code` folder of this lab.
   - Elaborate your codes and outputs as much as you can.
   - Try your best to answer the questions and complete the tasks and most importantly enjoy the process!!!

#### Import all the libraries that are necessary.

In [20]:
import pandas as pd
import numpy as np


#### Open the ` Employee.csv` file and name your dataset `employee`.

In [8]:
employee = pd.read_csv('Employee.csv')
employee


Name Department Education Gender      Title  Years  Salary
0    Jose         IT  Bachelor      M    analyst      1      35
1   Maria         IT    Master      F    analyst      2      30
2   David         HR    Master      M    analyst      2      30
3   Sonia         HR  Bachelor      F    analyst      4      35
4  Samuel      Sales    Master      M  associate      3      55
5     Eva      Sales  Bachelor      F  associate      2      55
6  Carlos         IT    Master      M         VP      8      70
7   Pedro         IT       Phd      M  associate      7      60
8     Ana         HR    Master      F         VP      8      70

#### What's the mean salary by department?
Hint : There are 2 ways to do it - one way is with groupby and the other way is with pivot tables. Do it both ways. 


In [10]:
employee.groupby(['Department']).agg({'Salary' : np.mean})


Salary
Department        
HR           45.00
IT           48.75
Sales        55.00

In [17]:
# your answer here
df_d=employee.pivot_table( index = ['Department'], values = ['Salary'], aggfunc= {'Salary': 'mean'}).reset_index()
df_d.columns = ['Department','Salary_Mean']
df_d

Department  Salary_Mean
0         HR        45.00
1         IT        48.75
2      Sales        55.00

#### We want now to group the data by more than one column. Use Pandas pivot table function to select the mean salary by department and title.

In [18]:

df_dt = employee.pivot_table( index = ['Department','Title'], values = ['Salary'], aggfunc= {'Salary': 'mean'}).reset_index()
df_dt.columns = ['Department','Title','Salary_Mean']
df_dt


Department      Title  Salary_Mean
0         HR         VP         70.0
1         HR    analyst         32.5
2         IT         VP         70.0
3         IT    analyst         32.5
4         IT  associate         60.0
5      Sales  associate         55.0

#### We want to display the number of years of work experience and the salary for each Title. Add Years as a column in your pivot table.
*Hint: Use Pandas doc(see References in README.md) to explore the pivot_table() arguments*

In [19]:
df_dty = employee.pivot_table( index = ['Department','Title','Years'], values = ['Salary'], aggfunc= {'Salary': 'mean'}).reset_index()
df_dty.columns = ['Department','Title','Years','Salary_Mean']
df_dty


Department      Title  Years  Salary_Mean
0         HR         VP      8           70
1         HR    analyst      2           30
2         HR    analyst      4           35
3         IT         VP      8           70
4         IT    analyst      1           35
5         IT    analyst      2           30
6         IT  associate      7           60
7      Sales  associate      2           55
8      Sales  associate      3           55

#### We want to know the information about the mean salary and number of employees in each department and for each job title.

In [24]:
df_det = employee.pivot_table(index = ['Department', 'Title'], values=['Salary'], aggfunc={'Salary':np.mean , 'Title': 'count'})
df_det                          


Salary  Title
Department Title                   
HR         VP           70.0      1
           analyst      32.5      2
IT         VP           70.0      1
           analyst      32.5      2
           associate    60.0      1
Sales      associate    55.0      2

#### We want to know for each department and for each job title the Salary's median, minimum, maximum and standard deviation. Comment your results.

In [29]:
df_detmmmd =employee.groupby(['Department','Title']).agg({'Salary': ['median','min','max','std']})
df_detmmmd    

Salary                  
                     median min max       std
Department Title                             
HR         VP          70.0  70  70       NaN
           analyst     32.5  30  35  3.535534
IT         VP          70.0  70  70       NaN
           analyst     32.5  30  35  3.535534
           associate   60.0  60  60       NaN
Sales      associate   55.0  55  55  0.000000

COMMENTS: 
for HR VP, the Median Salary: 70.0 Minimum Salary: 70 Maximum Salary: 70 and Standard Deviation is a NaN
This happens because  there is only 1 VP, so the median min and max are all the same and there is not standard deviation( we only have 1 value)
for HR Analyst, the Median Salary: 32.5, the Minimum Salary: 30 , the Maximum Salary: 35 and Standard Deviation (Salary): 3.535534
for IT VP we get only 70 for all results except standard deviation ( only 1 employee so min max and median are same and std does not exist)
for IT analyst, the Median Salary: 32.5, Minimum Salary: 30, Maximum Salary: 35 amd Standard Deviation (Salary): 3.535534
for IT associate, the values are all 60 except for std deviation ( same as it vp and hr vp) only 1 employee
for Sales associate the values are all 55, and std dev is 0, so there are 2 employees on this department but they receive the same salary I guess.

#### Based on your comments, fill the missing values with an appropriate value.

In [31]:
#I would replace all NaNs with 0, as there is no standard deviation for those job titles.
df_detmmmd = df_detmmmd.fillna(0)
df_detmmmd


Salary                  
                     median min max       std
Department Title                             
HR         VP          70.0  70  70  0.000000
           analyst     32.5  30  35  3.535534
IT         VP          70.0  70  70  0.000000
           analyst     32.5  30  35  3.535534
           associate   60.0  60  60  0.000000
Sales      associate   55.0  55  55  0.000000

#### The stake holders want to know for each department the number of employees and how much money is spend on salaries. Could you provide that information? 

In [32]:
df_dss = employee.pivot_table(index = ['Department'], values=['Salary'], aggfunc={'Salary':'sum'})
df_dss               

Salary
Department        
HR             135
IT             195
Sales          110

#### For each Department and Title, we want to know the the total years of work experience and the mean salary.

In [33]:
# I am assuming you want the total years of work combined of all employees.
df_detmmmd =employee.groupby(['Department','Title']).agg({'Years':'sum' ,'Salary': np.mean})
df_detmmmd    


Years  Salary
Department Title                   
HR         VP             8    70.0
           analyst        6    32.5
IT         VP             8    70.0
           analyst        3    32.5
           associate      7    60.0
Sales      associate      5    55.0

### Bonus

#### We now want to compute the mean salary after removing the maxima for each Department. Create the appropriate pivot table.

Hint: Write a custom function.

In [ ]:
# your answer here
